File: packages/java/src/main/java/ai/codec/LongestMatchTokenizer.java
// SPDX-License-Identifier: MIT
package ai.codec;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Vocab-only longest-prefix-match tokenizer. Walks input left-to-right,
 * emitting the ID of the longest vocab fragment that matches at each
 * position. Suitable for canonical-IR / synthetic test maps. NOT
 * BPE-correct for real model vocabs: use {@link BPETokenizer} for those.
 */
public final class LongestMatchTokenizer implements ITokenizer {
    private final String id;
    private final Map<String, Integer> fragmentToId = new HashMap<>();
    private final int maxFragmentLength;
    private final Map<String, Integer> specialFragmentToId = new HashMap<>();

    public LongestMatchTokenizer(TokenizerMap map) {
        this.id = map.id;
        int maxLen = 1;

        if (map.vocab != null) {
            for (Map.Entry<String, Integer> e : map.vocab.entrySet()) {
                String fragment = e.getKey();
                if (fragment == null || fragment.isEmpty()) continue;
                fragmentToId.put(fragment, e.getValue());
                if (fragment.length() > maxLen) maxLen = fragment.length();
            }
        }
        if (map.tokens != null) {
            for (Map.Entry<String, String> e : map.tokens.entrySet()) {
                String fragment = e.getValue();
                if (fragment == null || fragment.isEmpty()) continue;
                int parsedId;
                try { parsedId = Integer.parseInt(e.getKey()); }
                catch (NumberFormatException ex) { continue; }
                fragmentToId.put(fragment, parsedId);
                if (fragment.length() > maxLen) maxLen = fragment.length();
            }
        }
        this.maxFragmentLength = maxLen;

        if (map.specialTokens != null) {
            for (Map.Entry<String, Integer> e : map.specialTokens.entrySet()) {
                String name = e.getKey();
                int v = e.getValue();
                specialFragmentToId.put(name, v);
                if (!name.startsWith("<"))
                    specialFragmentToId.put("<|" + name + "|>", v);
            }
        }
    }

    @Override public String getId() { return id; }

    @Override
    public int[] encode(String text) {
        List<Integer> output = new ArrayList<>();
        int pos = 0;
        int n = text.length();

        while (pos < n) {
            // Specials win.
            boolean consumed = false;
            for (Map.Entry<String, Integer> e : specialFragmentToId.entrySet()) {
                String frag = e.getKey();
                if (text.regionMatches(pos, frag, 0, frag.length())) {
                    output.add(e.getValue());
                    pos += frag.length();
                    consumed = true;
                    break;
                }
            }
            if (consumed) continue;

            int remaining = n - pos;
            int tryUpTo = Math.min(maxFragmentLength, remaining);
            int matchedId = -1;
            int matchedLen = 0;
            for (int len = tryUpTo; len >= 1; len--) {
                String candidate = text.substring(pos, pos + len);
                Integer id = fragmentToId.get(candidate);
                if (id != null) {
                    matchedId = id;
                    matchedLen = len;
                    break;
                }
            }

            if (matchedId == -1) {
                output.add(0); // UNK
                pos += 1;
            } else {
                output.add(matchedId);
                pos += matchedLen;
            }
        }
        int[] arr = new int[output.size()];
        for (int i = 0; i < arr.length; i++) arr[i] = output.get(i);
        return arr;
    }
}

File: packages/dotnet/src/Codec.Net/LongestMatchTokenizer.cs
// SPDX-License-Identifier: MIT
namespace Codec;

/// <summary>
/// Vocab-only longest-prefix-match tokenizer. Walks input left-to-right,
/// emitting the ID of the longest vocab fragment that matches at each
/// position. Suitable for canonical-IR / synthetic test maps. NOT
/// BPE-correct for real model vocabs: use <see cref="BPETokenizer"/>
/// for those.
/// </summary>
public sealed class LongestMatchTokenizer : ITokenizer
{
    public string Id { get; }
    private readonly Dictionary<string, int> _fragmentToId = new();
    private readonly int _maxFragmentLength;
    private readonly Dictionary<string, int> _specialFragmentToId = new();

    public LongestMatchTokenizer(TokenizerMap map)
    {
        Id = map.Id;
        var maxLen = 1;

        if (map.Vocab is not null)
        {
            foreach (var (fragment, id) in map.Vocab)
            {
                if (string.IsNullOrEmpty(fragment)) continue;
                _fragmentToId[fragment] = id;
                if (fragment.Length > maxLen) maxLen = fragment.Length;
            }
        }
        if (map.Tokens is not null)
        {
            foreach (var (idStr, fragment) in map.Tokens)
            {
                if (string.IsNullOrEmpty(fragment)) continue;
                if (!int.TryParse(idStr, out var id)) continue;
                _fragmentToId[fragment] = id;
                if (fragment.Length > maxLen) maxLen = fragment.Length;
            }
        }
        _maxFragmentLength = maxLen;

        if (map.SpecialTokens is not null)
        {
            foreach (var (name, id) in map.SpecialTokens)
            {
                _specialFragmentToId[name] = id;
                if (!name.StartsWith('<'))
                    _specialFragmentToId[$"<|{name}|>"] = id;
            }
        }
    }

    public int[] Encode(string text)
    {
        var output = new List<int>();
        var pos = 0;
        var n = text.Length;

        while (pos < n)
        {
            // Specials win.
            var consumed = false;
            foreach (var (frag, id) in _specialFragmentToId)
            {
                if (string.CompareOrdinal(text, pos, frag, 0, frag.Length) == 0)
                {
                    output.Add(id);
                    pos += frag.Length;
                    consumed = true;
                    break;
                }
            }
            if (consumed) continue;

            var remaining = n - pos;
            var tryUpTo = Math.Min(_maxFragmentLength, remaining);
            var matchedId = -1;
            var matchedLen = 0;
            for (var len = tryUpTo; len >= 1; len--)
            {
                var candidate = text.Substring(pos, len);
                if (_fragmentToId.TryGetValue(candidate, out var id))
                {
                    matchedId = id;
                    matchedLen = len;
                    break;
                }
            }

            if (matchedId == -1)
            {
                output.Add(0); // UNK
                pos += 1;
            }
            else
            {
                output.Add(matchedId);
                pos += matchedLen;
            }
        }
        return output.ToArray();
    }
}

/// <summary>Top-level tokenizer factory.</summary>
public static class Tokenize
{
    /// <summary>
    /// Build the right tokenizer for the map. <see cref="BPETokenizer"/>
    /// when the map has BPE data; otherwise <see cref="LongestMatchTokenizer"/>.
    /// </summary>
    public static ITokenizer Pick(TokenizerMap map) =>
        BPETokenizer.Supports(map)
            ? new BPETokenizer(map)
            : new LongestMatchTokenizer(map);

    /// <summary>One-shot encode using <see cref="Pick"/>.</summary>
    public static int[] Encode(TokenizerMap map, string text) => Pick(map).Encode(text);
}
